experiments/imit_benchmark.sh

```diff
@@ -107,18 +105,12 @@ parallel -j 25% --header : --results "${LOG_ROOT}/parallel/" --colsep , --progre
   with \
   '{env_config_name}' seed='{seed}' \
   common.log_dir="${LOG_ROOT}/{env_config_name}_{seed}/n_expert_demos_{n_expert_demos}" \
-  demonstrations.data_dir="${DATA_DIR}" \
   demonstrations.n_expert_demos='{n_expert_demos}' \
   checkpoint_interval=0 \
   "${extra_configs[@]}" \
   :::: $CONFIG_CSV \
   ::: seed "${SEEDS[@]}"
 
-# Directory path is really long. Enter the directory to shorten results output,
-# which includes directory of each stdout file.
-pushd "${LOG_ROOT}/parallel"
```

Comment:
Apparently I removed it since it made some tests fail and I could not understand what it does. Right now I had another go at it but still could not figure out what those lines are good for. The comment seems to be unrelated and does not help in understanding. Do you have a hint for me @AdamGleave ?

What I did find out:
1. It makes `benchmark_and_table.sh --fast` fail (and therefore the experiment tests).
2. `benchmark_and_table.sh` fails on a call to `imit_benchmark.sh`
3. Just running `imit_benchmark.sh --fast` does not fail.
4. The line seems to ensure that certain result files or stdout files have been stored. In the case of the failing `benchmark_and_table.sh`, there is just one result file while `imit_benchmark.sh` itself generates two.

Reply:
In general please don't delete lines just because it makes the tests pass ;) Do feel free to ask me or anyone that shows up in git blame if the purpose of code is unclear. Sometimes it really should be deprecated, but often it's important.

I thought we discussed what these lines do back in https://github.com/HumanCompatibleAI/imitation/pull/445#issuecomment-1178374293 The TL;DR is that these lines print out a summary of the results from all the scripts being run. This is useful logging output. If the lines fail, it's because there's no results, which should never happen.

You can probably figure out what's going on by looking through the logs. Either the output format has changed, or some of the Python scripts are failing silently (if they returned non-zero exit code this should be passed through by parallel and fail earlier).

If you have trouble figuring this out let me know and I can take a look if you send me which tests are failing.

Reply:
You are right, deleting lines to make tests pass is not a very smart move :grimacing: . I also deleted it since I simply could not figure out what it is there fore. So I dug some deeper into the following line:

```bash
find . -name stdout -print0 | sort -z | xargs -0 tail -n 15 | grep -E '==|Result'
```

My best guess for it's intended action is: for each `stdout` file in the `parallel` subdir, check if it contains the string "Result" in the last 15 lines.
What I don't get is 1. why `sort`? and 2. why match for the string `==` OR for `Result`.
The sorting is meaningless but matching for `==` actually brings us into trouble later.

The first issue with this is that the `Result` seems to occur in `stderr` instead of `stdout`. Ok that is an easy fix; lets also remove the `sort`:

```bash
find . -name stderr -print0 | xargs -0 tail -n 15 | grep -E '==|Result'
```
Now as I understand the documentation of `find -print0` and `xargs -0`, the command executed in a directory like this:
```
.
└── env_config_name
    ├── seals_cartpole
    │   └── n_expert_demos
    │       └── 40
    │           └── seed
    │               └── 0
    │                   ├── seq
    │                   ├── stderr
    │                   └── stdout
    └── seals_mountain_car
        └── n_expert_demos
            └── 40
                └── seed
                    └── 0
                        ├── seq
                        ├── stderr
                        └── stdout
```
**should** result in two independent invocations of `tail`. What actually happens (you can verify this by enabling interactive mode of `xargs` with the `-p` flag) is that tail is executed once with the two files as an argument:
```bash
tail -n 15 ./env_config_name/seals_mountain_car/n_expert_demos/40/seed/0/stderr ./env_config_name/seals_cartpole/n_expert_demos/40/seed/0/stderr
```
This will output (with `-n 5` for brevity):
```
==> ./env_config_name/seals_mountain_car/n_expert_demos/40/seed/0/stderr <==
INFO - imitation.scripts.train_adversarial - Using '<class 'imitation.algorithms.adversarial.airl.AIRL'>' algorithm
round: 100%|██████████| 5/5 [00:00<00:00, 20.37it/s]
INFO - root - Saved policy to output/imit_benchmark/2022-08-26T12:42:38+02:00/seals_mountain_car_0/n_expert_demos_40/checkpoints/final/gen_policy
INFO - test-2022-08-26T12:42:24+02:00 - Result: {'imit_stats': {'n_traj': 2, 'monitor_return_len': 2, 'return_min': -4.088658452033997, 'return_mean': -1.3765016160905361, 'return_std': 2.7121568359434605, 'return_max': 1.3356552198529243, 'len_min': 5, 'len_mean': 5.0, 'len_std': 0.0, 'len_max': 5, 'monitor_return_min': -5.0, 'monitor_return_mean': -5.0, 'monitor_return_std': 0.0, 'monitor_return_max': -5.0}, 'expert_stats': {'n_traj': 40, 'return_min': -200.0, 'return_mean': -200.0, 'return_std': 0.0, 'return_max': -200.0, 'len_min': 200, 'len_mean': 200.0, 'len_std': 0.0, 'len_max': 200}}
INFO - test-2022-08-26T12:42:24+02:00 - Completed after 0:00:02

==> ./env_config_name/seals_cartpole/n_expert_demos/40/seed/0/stderr <==
INFO - imitation.scripts.train_adversarial - Using '<class 'imitation.algorithms.adversarial.airl.AIRL'>' algorithm
round: 100%|██████████| 5/5 [00:00<00:00, 20.15it/s]
INFO - root - Saved policy to output/imit_benchmark/2022-08-26T12:42:38+02:00/seals_cartpole_0/n_expert_demos_40/checkpoints/final/gen_policy
INFO - test-2022-08-26T12:42:24+02:00 - Result: {'imit_stats': {'n_traj': 2, 'monitor_return_len': 2, 'return_min': 0.24264731258153915, 'return_mean': 5.492057155817747, 'return_std': 5.249409843236208, 'return_max': 10.741466999053955, 'len_min': 5, 'len_mean': 5.0, 'len_std': 0.0, 'len_max': 5, 'monitor_return_min': 5.0, 'monitor_return_mean': 5.0, 'monitor_return_std': 0.0, 'monitor_return_max': 5.0}, 'expert_stats': {'n_traj': 40, 'return_min': 500.0, 'return_mean': 500.0, 'return_std': 0.0, 'return_max': 500.0, 'len_min': 500, 'len_mean': 500.0, 'len_std': 0.0, 'len_max': 500}}
INFO - test-2022-08-26T12:42:24+02:00 - Completed after 0:00:04
```
Note the lines like `==> ./env_config_name/seals_cartpole/n_expert_demos/40/seed/0/stderr <==`! They contain `==`! This is just `tail`, printing the file name surrounded by `==> FILENAME <==` as a header as soon as you pass multiple files.

So the current version of the command produces false positives due to two reasons:
1. As soon as more than one `stderr` file is found by `find`, `tail` produces the `==> FILENAME <==` headers, which are matched by `grep -E '==|Result'` no matter the content of the `stderr` files.
2. Even if we fixed 1. by changing to `grep -E 'Result'`, there is only one execution of of `grep`. So if only **one single** of the `stderr` files contains a `'Result'`, the test is positive.

Maybe I missed the intention of the bash command. Maybe there are ways to make it work in a meaningful way. But I think that is not the right way to go.
If one of the runs fails, that should be reflected in return code of `python -m imitation.scripts.train_adversarial` and the return code will be propagated by `parallel`.
If we **must** (for whatever reason) ensure, that successful training is indicated by a "Result" string in the `stdout` file, then that is a case for a unit test and not a bash one-liner.

So I propose:
1. To still remove the lines but add a unit test to check if there is `Result` in `stderr` iff the return code is 0.
2. Avoid setting up experiments as bash scripts in the future and minimize the number of bash scripts in the repo. They are hard to understand, easy to mess up and hard to debug.

What do you think @AdamGleave ?

Reply:
Thanks for the detailed analysis! This seems broadly correct. However, I think the point of these lines was not purely to check for successful execution (I agree we should check return code of parallel for that), but also to return to the user a convenient summary. That's the reason why `sort -z` was included (so that you get results in a consistent order, that's more user parseable), and why `==` was included in the `grep` (so you can see which file each one is coming from).

It sounds like Sacred changed where it prints Result from stdout to stderr, and we can fix this just by changing to `stderr`. I'd be inclined to do that, and add some comments documenting what this line is doing, since it was clearly easier to write than to read :roll_eyes:

I do agree we should eventually try to burn these bash files and replace them with Python, and that'll allow for much prettier presentation of results...

Reply:
Ok as a "printing the results" feature this totally makes sense. I added it back with the `stderr` fix and a comment, that hopefully makes it easier to understand in the future.